src/executor/hash_join.rs

```diff
@@ -13,89 +15,91 @@ pub struct HashJoinExecutor {
     pub left_child: BoxedExecutor,
     pub right_child: BoxedExecutor,
     pub join_op: BoundJoinOperator,
+    // TODO: filter by condition
     pub condition: BoundExpr,
     pub left_column_index: usize,
     pub right_column_index: usize,
-    pub data_types: Vec<DataType>,
+    pub left_types: Vec<DataType>,
+    pub right_types: Vec<DataType>,
 }
 
-// TODO : support other types of join: left/right/full join
 impl HashJoinExecutor {
-    fn execute_hash_join(
-        left_chunks: Vec<DataChunk>,
-        right_chunks: Vec<DataChunk>,
-        left_column_index: usize,
-        right_column_index: usize,
-        data_types: Vec<DataType>,
-    ) -> Result<Option<DataChunk>, ExecutorError> {
-        // Build hash table
+    #[try_stream(boxed, ok = DataChunk, error = ExecutorError)]
+    pub async fn execute(self) {
+        // collect all chunks from children
+        let left_chunks: Vec<DataChunk> = self.left_child.try_collect().await?;
+        let right_chunks: Vec<DataChunk> = self.right_child.try_collect().await?;
 
-        let mut hash_map: HashMap<DataValue, Vec<Vec<DataValue>>> = HashMap::new();
-        for left_chunk in &left_chunks {
-            for left_row_idx in 0..left_chunk.cardinality() {
-                let row = left_chunk.get_row_by_idx(left_row_idx);
-                let hash_data_value = &row[left_column_index];
+        // helper functions
+        let left_rows = || left_chunks.iter().flat_map(|chunk| chunk.rows());
+        let right_rows = || right_chunks.iter().flat_map(|chunk| chunk.rows());
 
-                hash_map
-                    .entry(hash_data_value.clone())
-                    .or_insert_with(Vec::new);
-                let val = hash_map.get_mut(hash_data_value).unwrap();
-                val.push(row);
-            }
+        // build
+        let mut hash_map: HashMap<DataValue, Vec<RowRef<'_>>> = HashMap::new();
+        for left_row in left_rows() {
+            let hash_value = left_row.get(self.left_column_index);
+            hash_map
+                .entry(hash_value)
+                .or_insert_with(Vec::new)
+                .push(left_row);
         }
-        let mut chunk_builders: Vec<ArrayBuilderImpl> = vec![];
-        for ty in &data_types {
-            chunk_builders.push(ArrayBuilderImpl::new(ty));
+
+        // probe
+        let mut builders = (self.left_types.iter())
+            .chain(self.right_types.iter())
+            .map(|ty| ArrayBuilderImpl::with_capacity(PROCESSING_WINDOW_SIZE, ty))
+            .collect_vec();
+        for right_row in right_rows() {
+            let hash_value = right_row.get(self.right_column_index);
+            for left_row in hash_map.get(&hash_value).unwrap_or(&vec![]) {
```

Comment:
I would prefer `unwrap_or_else` or create the empty vector once at the beginning of this function.

Reply:
`vec![]` (`Vec::new()`) is a const fn. So I guess the compiler should be able to optimize here?

Reply:
Get it! Then I think there's no need to optimize :)